sources/advanced/Command.ts

```diff
@@ -271,9 +296,38 @@ export abstract class Command<Context extends BaseContext = BaseContext> {
                 command.addRest({name: propertyName, required});
             });
 
-            this.registerTransformer(prototype, (state, command) => {
-                // @ts-ignore: The property is meant to have been defined by the child class
-                command[propertyName] = state.positionals.map(({value}) => value);
+            this.registerTransformer(prototype, (state, command, builder) => {
+                // The builder's arity.extra will always be NoLimits,
+                // because it is set when we call registerDefinition
+
+                // We can't splice while looping, so we push all indexes
+                // of positionals that should be removed into this array
+                const positionalsToRemove: number[] = [];
+
+                // We go over each positional
+                for (let i = 0; i < state.positionals.length; ++i) {
+                    const positional = state.positionals[i];
+
+                    const isNoLimits = positional.extra === NoLimits;
+                    const isLeading = positional.extra === false && i < builder.arity.leading.length;
+
+                    // We skip the positional if it isn't either:
+                    // 1) a NoLimits extra (i.e. an optional rest argument)
+                    // 2) a leading positional (i.e. a required rest argument)
+                    if (!isNoLimits && !isLeading)
+                        continue;
+
+                    // We mark this positional for removal
+                    positionalsToRemove.push(i);
+
+                    // We insert its value at the end of the array property.
+
+                    // @ts-ignore: The property is meant to have been defined by the child class
+                    (command[propertyName] ??= []).push(positional.value);
+                }
+
+                state.positionals = state.positionals
+                    .filter((positional, index) => !positionalsToRemove.includes(index))
```

Comment:
I don't understand 🤔 Why don't we just keep a single counter of the number of arguments that need to be inserted into the property, then slice it twice (once for what we keep, another time for what we reject)?

Reply:
> another time for what we reject

We shouldn't slice what we reject, as those arguments should be left alone so that future transformers can consume them instead. (Required to support `Command.String` after `Command.Rest`).

The way it currently works is that we go over all positionals and we only consume the ones that we need and we mark their indexes for removal so that we can filter them out after we loop. We leave alone all positionals that we don't consume.

> Why don't we just keep a single counter of the number of arguments that need to be inserted into the property, then slice it twice (once for what we keep, another time for what we reject)?

I'm not sure I fully understand your idea. 🤔

Reply:
I meant something like this:

```ts
let count = 0;
while (isRestPositional(state.positionals[count]))
  count += 1;

command[propertyName] = state.positionals.slice(0, count);
state.positionals = state.positionals.slice(count);
```

Writing it also makes me wonder why https://github.com/arcanis/clipanion/pull/42/files/0c7d939caa4ae72d4cc9c75dad6e32812e120958#diff-a2b947e619d41bc28ab12f6f3f8df8dbR318 is a `continue` rather than a `break` - is there any case where a rest list should skip an element in the middle (rather than just before or after)?

Reply:
> I meant something like this:

Thanks, I understand it now. I've pushed the change!

> Writing it also makes me wonder why arcanis/clipanion/pull/42/files/0c7d939caa4ae72d4cc9c75dad6e32812e120958#diff-a2b947e619d41bc28ab12f6f3f8df8dbR318 is a continue rather than a break - is there any case where a rest list should skip an element in the middle (rather than just before or after)?

Nope, it was a mistake on my part. Rest arguments can only be leading positionals (at the start) or `NoLimits` extras (in the middle).